### PR TITLE
Bonding script enhancement to choose creation of bonding  interface in Peer

### PR DESCRIPTION
**Pull request by hathyaga**
interface in peer

Also, bond_fail function is enhanced to consider various
failover conditions

Signed-off-by: Harsha Thyagaraja <harshkid@linux.vnet.ibm.com>

### CODE REVIEW

File: io/net/bonding.py [hunk 1]
```diff
@@ -45,10 +45,7 @@ def setUp(self):
         depends = []
         if detected_distro.name == "Ubuntu":
             depends.append("openssh-client")
-        # FIXME: "redhat" as the distro name for RHEL is deprecated
-        # on Avocado versions >= 50.0.  This is a temporary compatibility
-        # enabler for older runners, but should be removed soon
-        if detected_distro.name in ["rhel", "fedora", "centos", "redhat"]:
+        if detected_distro.name in ["redhat", "fedora", "centos"]:
```
Comment (narasimhan-v) [inline io/net/bonding.py]: Please do not delete this. Have it as it was before.

File: io/net/bonding.py [hunk 2]
```diff
@@ -110,13 +107,16 @@ def setUp(self):
             self.net_mask.append(mask)
         self.bonding_slave_file = "/sys/class/net/%s/bonding/slaves"\
                                   % self.bond_name
+        self.peer_bond_needed = self.params.get("peer_bond_needed",
+                                                default="False")
```
Comment (narasimhan-v) [inline io/net/bonding.py]: default=False

File: io/net/bonding.py [hunk 3]
```diff
@@ -165,20 +165,47 @@ def bond_fail(self, arg1):
...
+                cmd = "cat /proc/net/bonding/%s" % self.bond_name
+                process.system_output(cmd, shell=True, verbose=True)
+                cmd = "ifconfig %s up" % interface
+                time.sleep(10)
+                if process.system(cmd, shell=True, ignore_status=True) != 0:
+                    self.fail("bonding not working when trying to up the\
+                              interface %s" % interface)
+                time.sleep(5)
+            else:
+                self.fail("Need a min of 2 host interfaces to to test\
```
Comment (narasimhan-v) [inline io/net/bonding.py]: Should not fail here. Can just print a warn message or info

File: io/net/bonding.py [hunk 4]
```diff
@@ -165,20 +165,47 @@ def bond_fail(self, arg1):
...
+        process.system_output(cmd, shell=True, verbose=True)
+        if process.system(ping_cmd, shell=True, ignore_status=True) != 0:
+            self.log.info("Ping test failed which is expected ")
+        for interface in self.host_interfaces:
             cmd = "ifconfig %s up" % interface
+            time.sleep(10)
             if process.system(cmd, shell=True, ignore_status=True) != 0:
-                self.fail("bonding not working when trying to up the\
-                          interface %s" % interface)
+                self.log.info("Ping to the bonding interface failed which\
```
Comment (narasimhan-v) [inline io/net/bonding.py]: There is no ping tried here, but the info says ping failed

File: io/net/bonding.py.data/bonding.yaml [hunk 5]
```diff
@@ -18,3 +18,5 @@ peer_ip: "85.85.85.85"
 peer_interfaces: "eth3"
 bond_name: "bond13"
 user_name: "root"
+peer_bond_needed: False
+wait_time: "10"
```
Comment (narasimhan-v) [inline io/net/bonding.py.data/bonding.yaml]: This should ideally be used in all time.sleep()

File: io/net/bonding.py [hunk 6]
```diff
@@ -165,20 +165,47 @@ def bond_fail(self, arg1):
...
-            cmd = "cat /proc/net/bonding/%s" % self.bond_name
-            process.system_output(cmd, shell=True, verbose=True)
+                self.fail("Could not bring down the interface %s " % interface)
+            time.sleep(10)
+        ping_cmd = "ping -I %s %s -c 5"\
+            % (self.bond_name, self.peer_first_ipinterface)
+        cmd = "cat /proc/net/bonding/%s" % self.bond_name
+        process.system_output(cmd, shell=True, verbose=True)
+        if process.system(ping_cmd, shell=True, ignore_status=True) != 0:
+            self.log.info("Ping test failed which is expected ")
```
Comment (narasimhan-v) [inline io/net/bonding.py]: This should ideally use the `ping_check()`.

File: io/net/bonding.py.data/bonding.yaml (hunk 5 above)
Comment (hathyaga) [inline io/net/bonding.py.data/bonding.yaml]: can it be used only where it is needed? The intention to use this is to wait for the Bond interface to come up in the peer. If this is set to any large value, then the entire script will take a long time to finish

File: io/net/bonding.py (hunk 1 above)
Comment (vrbagalkote) [inline io/net/bonding.py]: Can all three 'if' be a part of if-else case?

File: io/net/bonding.py.data/bonding.yaml (hunk 5 above)
Comment (narasimhan-v) [inline io/net/bonding.py.data/bonding.yaml]: @hathyaga If it is for peer only, please indicate that. The parameter name can be `peer_wait_time`.
Regarding script taking long time, there are ways to implement this with wait time :-)

File: io/net/bonding.py (hunk 3 above)
Comment (vrbagalkote) [inline io/net/bonding.py]: @narasimhan-v If warn or info is given, test will still continue ahead right? If minimum number of interfaces are not given, test should not continue, So I fell fail is right. Plz let me know your thoughts.

Comment (narasimhan-v) [inline io/net/bonding.py]: Theoretically bonding will work even if it has one slave only.
So, we should save the failed case, and fail in the last.

**Comment by hathyaga** [comment]
Nrew PR having changes suggested: https://github.com/avocado-framework-tests/avocado-misc-tests/pull/484

Closing this PR.

**Pull request closed without merging**